docs/proposals/20210222-kubelet-authentication.md

```diff
@@ -0,0 +1,331 @@
+as appropriate.
+
+#### Core Specification
+- Core Cluster API MUST provide the following implementations of CSRs and signers:
+    - `cluster.x-k8s.io/kube-apiserver-client-kubelet-insecure` which implement an “Always Allow” type signer that provides equivalent security to Cluster API v1alpha3. This is only to be used for providers where no secure mechanism exists.
+    - `cluster.x-k8s.io/kube-apiserver-client-kubelet-tpm` and `cluster-x-k8s-io/kubelet-serving-tpm`
+        - Will implement TPM-based certificate signers and requesters based on the cloud-provider-gcp implementation.
+        - However, since the mechanism for retrieving endorsement keys varies across platforms, the TPM signer will additionally require a provider specific mechanism to provide the TPM Endorsement Key.
+
+#### Provider Specification
+
+##### All providers
+- All providers MUST insert a ProviderID within the KubernetesNodeProviderIdentifierOID extension attribute of the CSR
+- All signer names MUST be filled in by the provider’s controller in InfraCluster.Status.KubeletClientCertificateSigner and InfraCluster.Status.KubeletServingCertificateSigner
+- All providers SHOULD implement trust-on-first-use type mechanisms to prevent replay attacks
+
+##### Insecure providers
+- An insecure provider CANNOT implement certificate rotation or kubelet serving certificate signing
+- InfraCluster.Status.KubeletClientCertificateSigner MUST be set to cluster.x-k8s.io/kube-apiserver-client-kubelet-insecure
+- An insecure provider MUST use the cluster.x-k8s.io/kube-apiserver-client-kubelet-insecure signer
+
+##### Secure providers
+- A secure provider MUST implement certificate rotation and kubelet server certificate signing
+- A provider must register signers of:
+    - `cluster-x-k8s-io/kube-apiserver-client-kubelet-<provider>`
+    - `cluster-x-k8s-io/kubelet-serving-<provider>`
+- A secure provider MUST implement a secure attestation mechanism, based upon PEM-encoded blocks within the Certificate Signing Request.
+- Where a secure provider’s attestation mechanism does not include a nonce or timestamp to protect against replay attacks, the mechanism MUST implement a secondary time-limited attestation (e.g. AWS Instance Identity document + AWS HMACv4 signature).
+- A provider’s signer MUST run on the management cluster
+
+##### TPM based providers
+- A TPM provider MUST use the following certificate signers
+    - `cluster-x-k8s-io/kube-apiserver-client-kubelet-tpm`
+    - `cluster-x-k8s-io/kubelet-serving-tpm`
+- A TPM provider MUST annotate new CSRs as follows:
+    - Key: cluster-x-k8s-io/tpm-endorsement-key
+    - Value: Platform-specific endorsement key (e.g., retrieved from GCP Shielded VM API or VMware vCenter)
+
+#### Kubeadm
+Since this proposal essentially takes over part of the node registration process from kubeadm, we will require the following changes:
```

Comment:
Show we assume also an option for skipping token generation during kubeadm init?

Also, it is not fully cleat to me what is the expected coordination mechanism between kubeadm init/joint and the node bootstrapper. More specifically
- kubeadm init creates kubelet credentials for the first node in order to avoid a chicken egg problem. How are we envisioning the Kubelet serving certificate rotation daemon taking charge of those credentials/certificate rotation
- kubeadm join, when skipping the token TLS bootstrap, requires a kubeconfig file to access the cluster. How we make sure kubeadm waits for the credentials generated by the node attestation plugin?

Reply:
As discussed I'll add a section on how we're planning on using https://kubernetes.io/docs/reference/access-authn-authz/authentication/#configuration for kubelet's kubeconfig

Reply:
added